BeanInstantiationException caused by IllegalAccessError when running test since upgrading to 4.7.0

### Expected Behavior

Test, annotated with `@MicronautTest` fails when initialising context.  

```
java.lang.IllegalAccessError: class com.package.$MyClass$Introspection tried to access protected method 'void java.lang.Record.<init>()' (com.package.$MyClass$Introspection is in unnamed module of loader 'app'; java.lang.Record is in module java.base of loader 'bootstrap')
```

`MyClass` is annotated with `@Introspected`.

Problem was not occurring prior to upgrading to 4.7.0 (and Gradle plugin to 4.4.4).

### Actual Behaviour

```
io.micronaut.context.exceptions.BeanInstantiationException: Bean definition [com.package.RequestHandlerTest] could not be loaded: Error instantiating bean of type  [com.package.RequestHandlerTest]

Message: Failed to load a service: java.lang.IllegalAccessError: class com.package.$MyClass$Introspection tried to access protected method 'void java.lang.Record.<init>()' (com.package.$MyClass$Introspection is in unnamed module of loader 'app'; java.lang.Record is in module java.base of loader 'bootstrap')
Path Taken: new RequestHandlerTest()

	at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:2000)
	at io.micronaut.context.DefaultApplicationContext.initializeContext(DefaultApplicationContext.java:314)
	at io.micronaut.context.DefaultBeanContext.configureAndStartContext(DefaultBeanContext.java:3318)
	at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:345)
	at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:216)
	at io.micronaut.test.extensions.AbstractMicronautExtension.startApplicationContext(AbstractMicronautExtension.java:507)
	at io.micronaut.test.extensions.AbstractMicronautExtension.beforeClass(AbstractMicronautExtension.java:346)
	at io.micronaut.test.extensions.junit5.MicronautJunit5Extension.beforeAll(MicronautJunit5Extension.java:84)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: io.micronaut.context.exceptions.BeanInstantiationException: Error instantiating bean of type  [com.package.RequestHandlerTest]

Message: Failed to load a service: java.lang.IllegalAccessError: class com.package.$MyClass$Introspection tried to access protected method 'void java.lang.Record.<init>()' (com.package.$MyClass$Introspection is in unnamed module of loader 'app'; java.lang.Record is in module java.base of loader 'bootstrap')
Path Taken: new RequestHandlerTest()
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2349)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2304)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2316)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3127)
	at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:80)
	at io.micronaut.context.DefaultBeanContext.findOrCreateSingletonBeanRegistration(DefaultBeanContext.java:3029)
	at io.micronaut.context.DefaultBeanContext.initializeEagerBean(DefaultBeanContext.java:2702)
	at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:1994)
	... 8 more
Caused by: io.micronaut.core.io.service.SoftServiceLoader$ServiceLoadingException: Failed to load a service: java.lang.IllegalAccessError: class com.package.$MyClass$Introspection tried to access protected method 'void java.lang.Record.<init>()' (com.package.$MyClass$Introspection is in unnamed module of loader 'app'; java.lang.Record is in module java.base of loader 'bootstrap')
	at io.micronaut.core.io.service.ServiceScanner$ServiceInstanceLoader.collect(ServiceScanner.java:244)
	at io.micronaut.core.io.service.ServiceScanner$DefaultServiceCollector.collect(ServiceScanner.java:157)
	at io.micronaut.core.io.service.SoftServiceLoader.collectDynamicServices(SoftServiceLoader.java:199)
	at io.micronaut.core.io.service.SoftServiceLoader.collectAll(SoftServiceLoader.java:174)
	at io.micronaut.core.beans.DefaultBeanIntrospector.getIntrospections(DefaultBeanIntrospector.java:117)
	at io.micronaut.core.beans.DefaultBeanIntrospector.findIntrospection(DefaultBeanIntrospector.java:88)
	at io.micronaut.core.reflect.InstantiationUtils.tryInstantiate(InstantiationUtils.java:165)
	at io.micronaut.context.conditions.MatchesCustomCondition.matches(MatchesCustomCondition.java:58)
	at io.micronaut.context.AbstractInitializableBeanDefinitionAndReference.matches(AbstractInitializableBeanDefinitionAndReference.java:111)
	at io.micronaut.context.AbstractInitializableBeanDefinitionAndReference.isEnabled(AbstractInitializableBeanDefinitionAndReference.java:94)
	at io.micronaut.context.DefaultBeanContext$BeanDefinitionProducer.isDefinitionEnabled(DefaultBeanContext.java:4294)
	at io.micronaut.context.DefaultBeanContext$BeanDefinitionProducer.isDefinitionEnabled(DefaultBeanContext.java:4274)
	at io.micronaut.context.DefaultBeanContext.collectBeanCandidates(DefaultBeanContext.java:2167)
	at io.micronaut.context.DefaultBeanContext.findBeanCandidates(DefaultBeanContext.java:2135)
	at io.micronaut.context.DefaultBeanContext.findConcreteCandidateNoCache(DefaultBeanContext.java:3184)
	at io.micronaut.context.DefaultBeanContext.findConcreteCandidate(DefaultBeanContext.java:3168)
	at io.micronaut.context.DefaultBeanContext.findBeanDefinition(DefaultBeanContext.java:2777)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2746)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1745)
	at io.micronaut.context.AbstractBeanResolutionContext.getBean(AbstractBeanResolutionContext.java:89)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2188)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForField(AbstractInitializableBeanDefinition.java:1694)
	at com.package.$RequestHandlerTest$Definition.inject(Unknown Source)
	at com.package.$RequestHandlerTest$Definition.instantiate(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2334)
	... 15 more
Caused by: io.micronaut.core.io.service.SoftServiceLoader$ServiceLoadingException: java.lang.IllegalAccessError: class com.package.$MyClass$Introspection tried to access protected method 'void java.lang.Record.<init>()' (com.package.$MyClass$Introspection is in unnamed module of loader 'app'; java.lang.Record is in module java.base of loader 'bootstrap')
	at io.micronaut.core.io.service.SoftServiceLoader.lambda$collectDynamicServices$2(SoftServiceLoader.java:195)
	at io.micronaut.core.io.service.ServiceScanner$ServiceInstanceLoader.compute(ServiceScanner.java:235)
	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
	at java.base/java.util.concurrent.ForkJoinTask.doExec$$$capture(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java)
	at --- Async.Stack.Trace --- (captured by IntelliJ IDEA debugger)
	at java.base/java.util.concurrent.ForkJoinTask.fork(ForkJoinTask.java)
	at io.micronaut.core.io.service.ServiceScanner$DefaultServiceCollector.compute(ServiceScanner.java:135)
	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
	at java.base/java.util.concurrent.ForkJoinTask.doExec$$$capture(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
Caused by: java.lang.IllegalAccessError: class com.package.$MyClass$Introspection tried to access protected method 'void java.lang.Record.<init>()' (com.package.$MyClass$Introspection is in unnamed module of loader 'app'; java.lang.Record is in module java.base of loader 'bootstrap')
	at com.package.$MyClass$Introspection.<clinit>(Unknown Source)
	at java.base/jdk.internal.misc.Unsafe.allocateInstance(Native Method)
	at java.base/java.lang.invoke.DirectMethodHandle.allocateInstance(DirectMethodHandle.java:501)
	at io.micronaut.core.io.service.SoftServiceLoader.lambda$collectDynamicServices$2(SoftServiceLoader.java:187)
	... 14 more

```

### Steps To Reproduce

_No response_

### Environment Information

_No response_

### Example Application

_No response_

### Version

4.7.0

--- a/core-processor/src/main/java/io/micronaut/inject/beans/visitor/BeanIntrospectionWriter.java
+++ b/core-processor/src/main/java/io/micronaut/inject/beans/visitor/BeanIntrospectionWriter.java
@@ -722,7 +722,8 @@ final class BeanIntrospectionWriter extends AbstractClassFileWriter {
         dispatchWriter.buildGetTargetMethodByIndex(classWriter);
         buildFindIndexedProperty(classWriter);
         buildGetIndexedProperties(classWriter);
-        boolean hasBuilder = annotationMetadata != null && annotationMetadata.isPresent(Introspected.class, "builder");
+        boolean hasBuilder = annotationMetadata != null &&
+            (annotationMetadata.isPresent(Introspected.class, "builder") || annotationMetadata.hasDeclaredAnnotation("lombok.Builder"));
         if (defaultConstructor != null) {
             writeInstantiateMethod(classWriter, defaultConstructor, "instantiate");
             // in case invoked directly or via instantiateUnsafe

--- a/core-processor/src/main/java/io/micronaut/inject/beans/visitor/IntrospectedTypeElementVisitor.java
+++ b/core-processor/src/main/java/io/micronaut/inject/beans/visitor/IntrospectedTypeElementVisitor.java
@@ -67,6 +67,7 @@ public class IntrospectedTypeElementVisitor implements TypeElementVisitor<Object
      * The position of the visitor.
      */
     public static final int POSITION = -100;
+    private static final String ANN_LOMBOK_BUILDER = "lombok.Builder";
 
     private final Map<String, BeanIntrospectionWriter> writers = new LinkedHashMap<>(10);
 
@@ -180,62 +181,101 @@ public class IntrospectedTypeElementVisitor implements TypeElementVisitor<Object
             AnnotationClassValue<?> builderClass = builder.annotationClassValue("builderClass").orElse(null);
             String[] writePrefixes = builder.getAnnotation("accessorStyle", AccessorsStyle.class)
                 .map(a -> a.stringValues("writePrefixes")).orElse(new String[]{""});
-            if (builderMethod != null) {
-                MethodElement methodElement = element
-                    .getEnclosedElement(ElementQuery.ALL_METHODS.onlyStatic()
-                        .filter(m -> m.getName().equals(builderMethod) && !m.getGenericReturnType().isVoid())
-                        .onlyAccessible(element))
-                    .orElse(null);
-                if (methodElement != null) {
-                    ClassElement returnType = methodElement.getGenericReturnType();
-                    if (returnType.isPublic() || returnType.getPackageName().equals(element.getPackageName())) {
-                        AnnotationValueBuilder<Introspected> replaceIntrospected = AnnotationValue.builder(introspected, RetentionPolicy.RUNTIME);
-                        replaceIntrospected.member("builderClass", new AnnotationClassValue<>(returnType.getName()));
-                        element.annotate(replaceIntrospected.build());
-                        AnnotationMetadata methodMetadata = methodElement.getMethodAnnotationMetadata().getTargetAnnotationMetadata();
+            processBuilderDefinition(
+                element,
+                context,
+                introspected,
+                index,
+                targetPackage,
+                builderMethod,
+                creatorMethod,
+                writePrefixes,
+                builderClass
+            );
+        } else if (element.hasDeclaredAnnotation(ANN_LOMBOK_BUILDER)) {
+            AnnotationValue<Annotation> lombokBuilder = element.getAnnotation(ANN_LOMBOK_BUILDER);
+            String builderMethod = lombokBuilder.stringValue("builderMethodName").orElse("builder");
+            MethodElement methodElement = element
+                .getEnclosedElement(ElementQuery.ALL_METHODS.onlyStatic()
+                    .filter(m -> m.getName().equals(builderMethod) && !m.getGenericReturnType().isVoid())
+                    .onlyAccessible(element))
+                .orElse(null);
+            if (methodElement == null) {
+                // Lombok processing not done yet, try again in the next round.
+                throw new ElementPostponedToNextRoundException(element);
+            }
+            String creatorMethod = lombokBuilder.stringValue("buildMethodName").orElse("build");
+            String[] writePrefixes = lombokBuilder.stringValue("setterPrefix").map(sp -> new String[] { sp }).orElse(new String[]{""});
+            processBuilderDefinition(
+                element,
+                context,
+                introspected,
+                index,
+                targetPackage,
+                builderMethod,
+                creatorMethod,
+                writePrefixes,
+                null
+            );
+        }
+    }
 
-                        handleBuilder(
-                            element,
-                            context,
-                            creatorMethod,
-                            writePrefixes,
-                            methodElement,
-                            null,
-                            returnType,
-                            methodMetadata,
-                            index,
-                            targetPackage
-                        );
-                    } else {
-                        context.fail("Builder return type is not public. The method must be static and accessible.", methodElement);
-                    }
-                } else {
-                    context.fail("Method specified by builderMethod not found. The method must be static and accessible.", element);
-                }
-            } else if (builderClass != null) {
-                ClassElement builderClassElement = context.getClassElement(builderClass.getName()).orElse(null);
-                if (builderClassElement != null) {
+    private void processBuilderDefinition(ClassElement element, VisitorContext context, AnnotationValue<Introspected> introspected, int index, String targetPackage, String builderMethod, String creatorMethod, String[] writePrefixes, AnnotationClassValue<?> builderClass) {
+        if (builderMethod != null) {
+            MethodElement methodElement = element
+                .getEnclosedElement(ElementQuery.ALL_METHODS.onlyStatic()
+                    .filter(m -> m.getName().equals(builderMethod) && !m.getGenericReturnType().isVoid())
+                    .onlyAccessible(element))
+                .orElse(null);
+            if (methodElement != null) {
+                ClassElement returnType = methodElement.getGenericReturnType();
+                if (returnType.isPublic() || returnType.getPackageName().equals(element.getPackageName())) {
                     AnnotationValueBuilder<Introspected> replaceIntrospected = AnnotationValue.builder(introspected, RetentionPolicy.RUNTIME);
-                    replaceIntrospected.member("builderClass", new AnnotationClassValue<>(builderClassElement.getName()));
+                    replaceIntrospected.member("builderClass", new AnnotationClassValue<>(returnType.getName()));
                     element.annotate(replaceIntrospected.build());
+                    AnnotationMetadata methodMetadata = methodElement.getMethodAnnotationMetadata().getTargetAnnotationMetadata();
 
                     handleBuilder(
                         element,
                         context,
                         creatorMethod,
                         writePrefixes,
-                        builderClassElement.getPrimaryConstructor().orElse(null),
-                        builderClassElement.getDefaultConstructor().orElse(null),
-                        builderClassElement,
-                        builderClassElement.getTargetAnnotationMetadata(),
+                        methodElement,
+                        null,
+                        returnType,
+                        methodMetadata,
                         index,
-                        targetPackage);
+                        targetPackage
+                    );
                 } else {
-                    context.fail("Builder class not found on compilation classpath: " + builderClass.getName(), element);
+                    context.fail("Builder return type is not public. The method must be static and accessible.", methodElement);
                 }
             } else {
-                context.fail("When specifying the 'builder' member of @Introspected you must supply either a builderClass or builderMethod", element);
+                context.fail("Method " + builderMethod + "() specified by builderMethod not found. The method must be static and accessible.", element);
             }
+        } else if (builderClass != null) {
+            ClassElement builderClassElement = context.getClassElement(builderClass.getName()).orElse(null);
+            if (builderClassElement != null) {
+                AnnotationValueBuilder<Introspected> replaceIntrospected = AnnotationValue.builder(introspected, RetentionPolicy.RUNTIME);
+                replaceIntrospected.member("builderClass", new AnnotationClassValue<>(builderClassElement.getName()));
+                element.annotate(replaceIntrospected.build());
+
+                handleBuilder(
+                    element,
+                    context,
+                    creatorMethod,
+                    writePrefixes,
+                    builderClassElement.getPrimaryConstructor().orElse(null),
+                    builderClassElement.getDefaultConstructor().orElse(null),
+                    builderClassElement,
+                    builderClassElement.getTargetAnnotationMetadata(),
+                    index,
+                    targetPackage);
+            } else {
+                context.fail("Builder class not found on compilation classpath: " + builderClass.getName(), element);
+            }
+        } else {
+            context.fail("When specifying the 'builder' member of @Introspected you must supply either a builderClass or builderMethod", element);
         }
     }
 
@@ -380,13 +420,18 @@ public class IntrospectedTypeElementVisitor implements TypeElementVisitor<Object
         List<PropertyElement> beanProperties = ce.getBeanProperties(propertyElementQuery).stream()
             .filter(p -> !p.isExcluded())
             .toList();
-        Optional<MethodElement> constructorElement = ce.getPrimaryConstructor();
-        constructorElement.ifPresent(constructorEl -> {
-            if (ArrayUtils.isNotEmpty(constructorEl.getParameters())) {
-                writer.visitConstructor(constructorEl);
-            }
-        });
-        ce.getDefaultConstructor().ifPresent(writer::visitDefaultConstructor);
+        // unfortunately sometimes we don't see the Lombok transformations
+        // so assume if the class is annotated with Lombok builder we cannot
+        // access the constructor.
+        if (!ce.hasDeclaredAnnotation(ANN_LOMBOK_BUILDER)) {
+            Optional<MethodElement> constructorElement = ce.getPrimaryConstructor();
+            constructorElement.ifPresent(constructorEl -> {
+                if (ArrayUtils.isNotEmpty(constructorEl.getParameters())) {
+                    writer.visitConstructor(constructorEl);
+                }
+            });
+            ce.getDefaultConstructor().ifPresent(writer::visitDefaultConstructor);
+        }
 
         for (PropertyElement beanProperty : beanProperties) {
             if (beanProperty.isExcluded()) {

--- a/inject-java/src/main/java/io/micronaut/annotation/processing/TypeElementVisitorProcessor.java
+++ b/inject-java/src/main/java/io/micronaut/annotation/processing/TypeElementVisitorProcessor.java
@@ -35,6 +35,7 @@ import io.micronaut.inject.ast.FieldElement;
 import io.micronaut.inject.ast.MemberElement;
 import io.micronaut.inject.ast.MethodElement;
 import io.micronaut.inject.processing.ProcessingException;
+import io.micronaut.inject.visitor.ElementPostponedToNextRoundException;
 import io.micronaut.inject.visitor.TypeElementVisitor;
 import io.micronaut.inject.visitor.VisitorContext;
 import io.micronaut.inject.writer.AbstractBeanDefinitionBuilder;
@@ -42,6 +43,7 @@ import io.micronaut.inject.writer.AbstractBeanDefinitionBuilder;
 import javax.annotation.processing.ProcessingEnvironment;
 import javax.annotation.processing.RoundEnvironment;
 import javax.annotation.processing.SupportedOptions;
+import javax.lang.model.element.Element;
 import javax.lang.model.element.TypeElement;
 import javax.lang.model.type.TypeMirror;
 import java.io.IOException;
@@ -276,6 +278,15 @@ public class TypeElementVisitorProcessor extends AbstractInjectAnnotationProcess
                             error(originatingElement.element(), e.getMessage());
                         } catch (PostponeToNextRoundException e) {
                             postponedTypes.put(javaClassElement.getCanonicalName(), e.getErrorElement());
+                        } catch (ElementPostponedToNextRoundException e) {
+                            Object nativeType = e.getOriginatingElement().getNativeType();
+                            if (nativeType instanceof JavaNativeElement jne) {
+                                Element element = jne.element();
+                                postponedTypes.put(javaClassElement.getCanonicalName(), element);
+                            } else {
+                                // should never happen.
+                                throw e;
+                            }
                         }
                     }
                 }

--- a/inject/src/main/java/io/micronaut/inject/beans/AbstractInitializableBeanIntrospection.java
+++ b/inject/src/main/java/io/micronaut/inject/beans/AbstractInitializableBeanIntrospection.java
@@ -91,11 +91,11 @@ public abstract class AbstractInitializableBeanIntrospection<B> implements Unsaf
     private IntrospectionBuilderData builderData;
 
     protected AbstractInitializableBeanIntrospection(Class<B> beanType,
-                                                     AnnotationMetadata annotationMetadata,
-                                                     AnnotationMetadata constructorAnnotationMetadata,
-                                                     Argument<?>[] constructorArguments,
-                                                     BeanPropertyRef<Object>[] propertiesRefs,
-                                                     BeanMethodRef<Object>[] methodsRefs) {
+                                                  AnnotationMetadata annotationMetadata,
+                                                  AnnotationMetadata constructorAnnotationMetadata,
+                                                  Argument<?>[] constructorArguments,
+                                                  BeanPropertyRef<Object>[] propertiesRefs,
+                                                  BeanMethodRef<Object>[] methodsRefs) {
         this.beanType = beanType;
         this.annotationMetadata = annotationMetadata == null ? AnnotationMetadata.EMPTY_METADATA : EvaluatedAnnotationMetadata.wrapIfNecessary(annotationMetadata);
         this.constructorAnnotationMetadata = constructorAnnotationMetadata == null ? AnnotationMetadata.EMPTY_METADATA : EvaluatedAnnotationMetadata.wrapIfNecessary(constructorAnnotationMetadata);
@@ -265,7 +265,7 @@ public abstract class AbstractInitializableBeanIntrospection<B> implements Unsaf
      * @param index  The method index
      * @param target The target
      * @param arg    The argument
-     * @param <V>    The result type
+     * @param <V> The result type
      * @return The result
      */
     @Nullable
@@ -474,63 +474,65 @@ public abstract class AbstractInitializableBeanIntrospection<B> implements Unsaf
                 .flatMap(a -> a.getAnnotation("builder", Introspected.IntrospectionBuilder.class)).orElse(null);
             Class<?> builderClass = getAnnotationMetadata().classValue(Introspected.class, "builderClass").orElse(null);
             if (builderAnn != null || builderClass != null) {
-                BeanIntrospection<Object> builderIntrospection = (BeanIntrospection<Object>) BeanIntrospection.getIntrospection(builderClass);
-                Collection<BeanMethod<Object, Object>> beanMethods = builderIntrospection.getBeanMethods();
-
-                // find the creator method
-                BeanMethod<Object, Object> constructorMethod = beanMethods.stream()
-                    .filter(m -> m.getReturnType().getType().equals(getBeanType()))
-                    .findFirst().orElse(null);
-                if (constructorMethod == null) {
-                    throw new IntrospectionException("No build method found in builder: " + builderClass.getName());
+                if (builderClass == null) {
+                    throw new IntrospectionException("Introspection defines invalid builder member for type: " + getBeanType());
                 } else {
-                    BeanMethod<Object, Object>[] builderMethods = beanMethods.stream()
-                        .filter(m ->
-                            m.getReturnType().getType().isAssignableFrom(builderIntrospection.getBeanType())
-                        )
-                        .toArray(BeanMethod[]::new);
+                    BeanIntrospection<Object> builderIntrospection = (BeanIntrospection<Object>) BeanIntrospection.getIntrospection(builderClass);
+                    Collection<BeanMethod<Object, Object>> beanMethods = builderIntrospection.getBeanMethods();
 
-                    List<Argument<?>> arguments = new ArrayList<>(builderMethods.length);
-                    Set<String> properties = CollectionUtils.newHashSet(builderMethods.length);
-                    Set<BeanMethod<?, ?>> excludedMethods = new HashSet<>();
-                    for (BeanMethod<Object, Object> builderMethod : builderMethods) {
-                        Argument<?> argument;
-                        @NonNull Argument<?>[] methodArgs = builderMethod.getArguments();
-                        if (ArrayUtils.isNotEmpty(methodArgs)) {
-                            argument = toWrapperIfNecessary(methodArgs[0]);
-                        } else {
-                            argument = Argument.of(Boolean.class, builderMethod.getName());
-                        }
-                        if (!properties.add(argument.getName())) {
-                            excludedMethods.add(builderMethod);
-                        } else {
-                            arguments.add(argument);
-                        }
-                    }
-                    if (!excludedMethods.isEmpty()) {
-                        builderMethods = Arrays.stream(builderMethods)
-                            .filter(bm -> !excludedMethods.contains(bm))
+                    // find the creator method
+                    BeanMethod<Object, Object> constructorMethod = beanMethods.stream()
+                        .filter(m -> m.getReturnType().getType().equals(getBeanType()))
+                        .findFirst().orElse(null);
+                    if (constructorMethod == null) {
+                        throw new IntrospectionException("No build method found in builder: " + builderClass.getName());
+                    } else {
+                        BeanMethod<Object, Object>[] builderMethods = beanMethods.stream()
+                            .filter(m ->
+                                m.getReturnType().getType().isAssignableFrom(builderIntrospection.getBeanType())
+                            )
                             .toArray(BeanMethod[]::new);
+
+                        List<Argument<?>> arguments = new ArrayList<>(builderMethods.length);
+                        Set<String> properties = CollectionUtils.newHashSet(builderMethods.length);
+                        Set<BeanMethod<?, ?>> excludedMethods = new HashSet<>();
+                        for (BeanMethod<Object, Object> builderMethod : builderMethods) {
+                            Argument<?> argument;
+                            @NonNull Argument<?>[] methodArgs = builderMethod.getArguments();
+                            if (ArrayUtils.isNotEmpty(methodArgs)) {
+                                argument = toWrapperIfNecessary(methodArgs[0]);
+                            } else {
+                                argument = Argument.of(Boolean.class, builderMethod.getName());
+                            }
+                            if (!properties.add(argument.getName())) {
+                                excludedMethods.add(builderMethod);
+                            } else {
+                                arguments.add(argument);
+                            }
+                        }
+                        if (!excludedMethods.isEmpty()) {
+                            builderMethods = Arrays.stream(builderMethods)
+                                .filter(bm -> !excludedMethods.contains(bm))
+                                .toArray(BeanMethod[]::new);
+                        }
+                        this.builderData = new IntrospectionBuilderData(
+                            builderIntrospection,
+                            constructorMethod,
+                            builderMethods,
+                            arguments.toArray(Argument.ZERO_ARGUMENTS)
+                        );
                     }
-                    this.builderData = new IntrospectionBuilderData(
-                        builderIntrospection,
-                        constructorMethod,
-                        builderMethods,
-                        arguments.toArray(Argument.ZERO_ARGUMENTS)
-                    );
                 }
             } else {
-                throw new IntrospectionException("Introspection defines invalid builder member for type: " + getBeanType());
-            }
-        } else {
-            int constructorLength = constructorArguments.length;
-            @NonNull UnsafeBeanProperty<B, Object>[] writeableProperties = resolveWriteableProperties(beanPropertiesList);
+                int constructorLength = constructorArguments.length;
+                @NonNull UnsafeBeanProperty<B, Object>[] writeableProperties = resolveWriteableProperties(beanPropertiesList);
 
-            this.builderData = new IntrospectionBuilderData(
-                constructorArguments,
-                constructorLength,
-                (UnsafeBeanProperty<Object, Object>[]) writeableProperties
-            );
+                this.builderData = new IntrospectionBuilderData(
+                    constructorArguments,
+                    constructorLength,
+                    (UnsafeBeanProperty<Object, Object>[]) writeableProperties
+                );
+            }
         }
         return builderData;
     }
@@ -667,7 +669,7 @@ public abstract class AbstractInitializableBeanIntrospection<B> implements Unsaf
     }
 
     private static final class IntrospectionBuilder<B> implements Builder<B> {
-        private static final Object[] NULL_ARG = {null};
+        private static final Object[] NULL_ARG = { null };
         private final Object[] params;
         private final IntrospectionBuilderData builderData;
         private final AbstractInitializableBeanIntrospection<B> introspection;
@@ -990,10 +992,10 @@ public abstract class AbstractInitializableBeanIntrospection<B> implements Unsaf
         @Override
         public String toString() {
             return "BeanProperty{" +
-                "beanType=" + beanType +
-                ", type=" + ref.argument.getType() +
-                ", name='" + ref.argument.getName() + '\'' +
-                '}';
+                    "beanType=" + beanType +
+                    ", type=" + ref.argument.getType() +
+                    ", name='" + ref.argument.getName() + '\'' +
+                    '}';
         }
     }
 
@@ -1087,10 +1089,10 @@ public abstract class AbstractInitializableBeanIntrospection<B> implements Unsaf
         @Override
         public String toString() {
             return "BeanWriteProperty{" +
-                "beanType=" + beanType +
-                ", type=" + argument.getType() +
-                ", name='" + argument.getName() + '\'' +
-                '}';
+                    "beanType=" + beanType +
+                    ", type=" + argument.getType() +
+                    ", name='" + argument.getName() + '\'' +
+                    '}';
         }
     }
 
@@ -1158,10 +1160,10 @@ public abstract class AbstractInitializableBeanIntrospection<B> implements Unsaf
         @Override
         public String toString() {
             return "BeanReadProperty{" +
-                "beanType=" + beanType +
-                ", type=" + argument.getType() +
-                ", name='" + argument.getName() + '\'' +
-                '}';
+                    "beanType=" + beanType +
+                    ", type=" + argument.getType() +
+                    ", name='" + argument.getName() + '\'' +
+                    '}';
         }
     }
 

--- a/inject/src/main/java/io/micronaut/inject/beans/AbstractInitializableBeanIntrospection.java
+++ b/inject/src/main/java/io/micronaut/inject/beans/AbstractInitializableBeanIntrospection.java
@@ -91,11 +91,11 @@ public abstract class AbstractInitializableBeanIntrospection<B> implements Unsaf
     private IntrospectionBuilderData builderData;
 
     protected AbstractInitializableBeanIntrospection(Class<B> beanType,
-                                                  AnnotationMetadata annotationMetadata,
-                                                  AnnotationMetadata constructorAnnotationMetadata,
-                                                  Argument<?>[] constructorArguments,
-                                                  BeanPropertyRef<Object>[] propertiesRefs,
-                                                  BeanMethodRef<Object>[] methodsRefs) {
+                                                     AnnotationMetadata annotationMetadata,
+                                                     AnnotationMetadata constructorAnnotationMetadata,
+                                                     Argument<?>[] constructorArguments,
+                                                     BeanPropertyRef<Object>[] propertiesRefs,
+                                                     BeanMethodRef<Object>[] methodsRefs) {
         this.beanType = beanType;
         this.annotationMetadata = annotationMetadata == null ? AnnotationMetadata.EMPTY_METADATA : EvaluatedAnnotationMetadata.wrapIfNecessary(annotationMetadata);
         this.constructorAnnotationMetadata = constructorAnnotationMetadata == null ? AnnotationMetadata.EMPTY_METADATA : EvaluatedAnnotationMetadata.wrapIfNecessary(constructorAnnotationMetadata);
@@ -265,7 +265,7 @@ public abstract class AbstractInitializableBeanIntrospection<B> implements Unsaf
      * @param index  The method index
      * @param target The target
      * @param arg    The argument
-     * @param <V> The result type
+     * @param <V>    The result type
      * @return The result
      */
     @Nullable
@@ -472,67 +472,65 @@ public abstract class AbstractInitializableBeanIntrospection<B> implements Unsaf
 
             AnnotationValue<Introspected.IntrospectionBuilder> builderAnn = getAnnotationMetadata().findAnnotation(Introspected.class)
                 .flatMap(a -> a.getAnnotation("builder", Introspected.IntrospectionBuilder.class)).orElse(null);
-            if (builderAnn != null) {
-                Class<?> builderClass = getAnnotationMetadata().classValue(Introspected.class, "builderClass").orElse(null);
-                if (builderClass != null) {
-                    BeanIntrospection<Object> builderIntrospection = (BeanIntrospection<Object>) BeanIntrospection.getIntrospection(builderClass);
-                    Collection<BeanMethod<Object, Object>> beanMethods = builderIntrospection.getBeanMethods();
+            Class<?> builderClass = getAnnotationMetadata().classValue(Introspected.class, "builderClass").orElse(null);
+            if (builderAnn != null || builderClass != null) {
+                BeanIntrospection<Object> builderIntrospection = (BeanIntrospection<Object>) BeanIntrospection.getIntrospection(builderClass);
+                Collection<BeanMethod<Object, Object>> beanMethods = builderIntrospection.getBeanMethods();
 
-                    // find the creator method
-                    BeanMethod<Object, Object> constructorMethod = beanMethods.stream()
-                        .filter(m -> m.getReturnType().getType().equals(getBeanType()))
-                        .findFirst().orElse(null);
-                    if (constructorMethod == null) {
-                        throw new IntrospectionException("No build method found in builder: " + builderClass.getName());
-                    } else {
-                        BeanMethod<Object, Object>[] builderMethods = beanMethods.stream()
-                            .filter(m ->
-                                m.getReturnType().getType().isAssignableFrom(builderIntrospection.getBeanType())
-                            )
-                            .toArray(BeanMethod[]::new);
-
-                        List<Argument<?>> arguments = new ArrayList<>(builderMethods.length);
-                        Set<String> properties = CollectionUtils.newHashSet(builderMethods.length);
-                        Set<BeanMethod<?, ?>> excludedMethods = new HashSet<>();
-                        for (BeanMethod<Object, Object> builderMethod : builderMethods) {
-                            Argument<?> argument;
-                            @NonNull Argument<?>[] methodArgs = builderMethod.getArguments();
-                            if (ArrayUtils.isNotEmpty(methodArgs)) {
-                                argument = toWrapperIfNecessary(methodArgs[0]);
-                            } else {
-                                argument = Argument.of(Boolean.class, builderMethod.getName());
-                            }
-                            if (!properties.add(argument.getName())) {
-                                excludedMethods.add(builderMethod);
-                            } else {
-                                arguments.add(argument);
-                            }
-                        }
-                        if (!excludedMethods.isEmpty()) {
-                            builderMethods = Arrays.stream(builderMethods)
-                                .filter(bm -> !excludedMethods.contains(bm))
-                                .toArray(BeanMethod[]::new);
-                        }
-                        this.builderData = new IntrospectionBuilderData(
-                            builderIntrospection,
-                            constructorMethod,
-                            builderMethods,
-                            arguments.toArray(Argument.ZERO_ARGUMENTS)
-                        );
-                    }
+                // find the creator method
+                BeanMethod<Object, Object> constructorMethod = beanMethods.stream()
+                    .filter(m -> m.getReturnType().getType().equals(getBeanType()))
+                    .findFirst().orElse(null);
+                if (constructorMethod == null) {
+                    throw new IntrospectionException("No build method found in builder: " + builderClass.getName());
                 } else {
-                    throw new IntrospectionException("Introspection defines invalid builder member for type: " + getBeanType());
+                    BeanMethod<Object, Object>[] builderMethods = beanMethods.stream()
+                        .filter(m ->
+                            m.getReturnType().getType().isAssignableFrom(builderIntrospection.getBeanType())
+                        )
+                        .toArray(BeanMethod[]::new);
+
+                    List<Argument<?>> arguments = new ArrayList<>(builderMethods.length);
+                    Set<String> properties = CollectionUtils.newHashSet(builderMethods.length);
+                    Set<BeanMethod<?, ?>> excludedMethods = new HashSet<>();
+                    for (BeanMethod<Object, Object> builderMethod : builderMethods) {
+                        Argument<?> argument;
+                        @NonNull Argument<?>[] methodArgs = builderMethod.getArguments();
+                        if (ArrayUtils.isNotEmpty(methodArgs)) {
+                            argument = toWrapperIfNecessary(methodArgs[0]);
+                        } else {
+                            argument = Argument.of(Boolean.class, builderMethod.getName());
+                        }
+                        if (!properties.add(argument.getName())) {
+                            excludedMethods.add(builderMethod);
+                        } else {
+                            arguments.add(argument);
+                        }
+                    }
+                    if (!excludedMethods.isEmpty()) {
+                        builderMethods = Arrays.stream(builderMethods)
+                            .filter(bm -> !excludedMethods.contains(bm))
+                            .toArray(BeanMethod[]::new);
+                    }
+                    this.builderData = new IntrospectionBuilderData(
+                        builderIntrospection,
+                        constructorMethod,
+                        builderMethods,
+                        arguments.toArray(Argument.ZERO_ARGUMENTS)
+                    );
                 }
             } else {
-                int constructorLength = constructorArguments.length;
-                @NonNull UnsafeBeanProperty<B, Object>[] writeableProperties = resolveWriteableProperties(beanPropertiesList);
-
-                this.builderData = new IntrospectionBuilderData(
-                    constructorArguments,
-                    constructorLength,
-                    (UnsafeBeanProperty<Object, Object>[]) writeableProperties
-                );
+                throw new IntrospectionException("Introspection defines invalid builder member for type: " + getBeanType());
             }
+        } else {
+            int constructorLength = constructorArguments.length;
+            @NonNull UnsafeBeanProperty<B, Object>[] writeableProperties = resolveWriteableProperties(beanPropertiesList);
+
+            this.builderData = new IntrospectionBuilderData(
+                constructorArguments,
+                constructorLength,
+                (UnsafeBeanProperty<Object, Object>[]) writeableProperties
+            );
         }
         return builderData;
     }
@@ -669,7 +667,7 @@ public abstract class AbstractInitializableBeanIntrospection<B> implements Unsaf
     }
 
     private static final class IntrospectionBuilder<B> implements Builder<B> {
-        private static final Object[] NULL_ARG = { null };
+        private static final Object[] NULL_ARG = {null};
         private final Object[] params;
         private final IntrospectionBuilderData builderData;
         private final AbstractInitializableBeanIntrospection<B> introspection;
@@ -992,10 +990,10 @@ public abstract class AbstractInitializableBeanIntrospection<B> implements Unsaf
         @Override
         public String toString() {
             return "BeanProperty{" +
-                    "beanType=" + beanType +
-                    ", type=" + ref.argument.getType() +
-                    ", name='" + ref.argument.getName() + '\'' +
-                    '}';
+                "beanType=" + beanType +
+                ", type=" + ref.argument.getType() +
+                ", name='" + ref.argument.getName() + '\'' +
+                '}';
         }
     }
 
@@ -1089,10 +1087,10 @@ public abstract class AbstractInitializableBeanIntrospection<B> implements Unsaf
         @Override
         public String toString() {
             return "BeanWriteProperty{" +
-                    "beanType=" + beanType +
-                    ", type=" + argument.getType() +
-                    ", name='" + argument.getName() + '\'' +
-                    '}';
+                "beanType=" + beanType +
+                ", type=" + argument.getType() +
+                ", name='" + argument.getName() + '\'' +
+                '}';
         }
     }
 
@@ -1160,10 +1158,10 @@ public abstract class AbstractInitializableBeanIntrospection<B> implements Unsaf
         @Override
         public String toString() {
             return "BeanReadProperty{" +
-                    "beanType=" + beanType +
-                    ", type=" + argument.getType() +
-                    ", name='" + argument.getName() + '\'' +
-                    '}';
+                "beanType=" + beanType +
+                ", type=" + argument.getType() +
+                ", name='" + argument.getName() + '\'' +
+                '}';
         }
     }
 

--- a/test-suite/build.gradle
+++ b/test-suite/build.gradle
@@ -118,3 +118,4 @@ test {
     // Prevent scanning classes with missing classes
     exclude '**/classnotfound/**'
 }
+

--- a/test-suite/src/test/java/io/micronaut/test/lombok/LombokIntrospectedBuilderTest.java
+++ b/test-suite/src/test/java/io/micronaut/test/lombok/LombokIntrospectedBuilderTest.java
@@ -1,5 +1,7 @@
 package io.micronaut.test.lombok;
 
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
 import io.micronaut.core.beans.BeanIntrospection;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
@@ -17,7 +19,19 @@ public class LombokIntrospectedBuilderTest {
         RobotEntity robotEntity = builder.with("name", "foo")
             .build();
 
-        Assertions.assertEquals("foo", robotEntity.getName());
+        assertEquals("foo", robotEntity.getName());
+    }
+
+    @Test
+    void testLombokBuilder2() {
+        BeanIntrospection.Builder<MyEntity> builder = BeanIntrospection.getIntrospection(MyEntity.class)
+            .builder();
+        MyEntity.MyEntityBuilder builder1 = MyEntity.builder();
+        builder.with("name", "foo");
+        builder.with("id", "123");
+        MyEntity myEntity = builder.build();
+        assertEquals("foo", myEntity.getName());
+        assertEquals("123", myEntity.getId());
     }
 
     @Test
@@ -29,7 +43,7 @@ public class LombokIntrospectedBuilderTest {
         SimpleEntity simpleEntity = builder.with("id", id)
             .build();
 
-        Assertions.assertEquals(id, simpleEntity.getId());
+        assertEquals(id, simpleEntity.getId());
 
         BeanIntrospection<SimpleEntity.CompartmentCreationTimeIndexPrefix> innerClassIntrospection =
             BeanIntrospection.getIntrospection(SimpleEntity.CompartmentCreationTimeIndexPrefix.class);
@@ -42,7 +56,7 @@ public class LombokIntrospectedBuilderTest {
         SimpleEntity.CompartmentCreationTimeIndexPrefix innerClassEntity =
             innerClassBuilder.with("compartmentId", "c1").with("timeCreated", current).build();
 
-        Assertions.assertEquals("c1", innerClassEntity.getCompartmentId());
-        Assertions.assertEquals(current, innerClassEntity.getTimeCreated());
+        assertEquals("c1", innerClassEntity.getCompartmentId());
+        assertEquals(current, innerClassEntity.getTimeCreated());
     }
 }

--- a/test-suite/src/test/java/io/micronaut/test/lombok/MyEntity.java
+++ b/test-suite/src/test/java/io/micronaut/test/lombok/MyEntity.java
@@ -1,0 +1,19 @@
+package io.micronaut.test.lombok;
+
+
+import io.micronaut.core.annotation.Introspected;
+import lombok.Builder;
+import lombok.Value;
+
+@Introspected
+@Value
+@Builder
+public class MyEntity {
+    public static final String NAME_INDEX = "name";
+
+    @lombok.NonNull
+    String id;
+
+    @lombok.NonNull
+    String name;
+}

